#  Data Analysis using Pandas and NumPy Lesson
> A series of lessons based on the utilization of Pandas and NumPy to read and create interesting things with files and data. 
- toc: true
- categories: [t3lessons]
- week: 31

# Predictive Analysis
Predictive analysis is the use of statistical, data mining, and machine learning techniques to analyze current and historical data in order to make predictions about future events or behaviors. It involves identifying patterms and trends in data, and then using that information to forecast what is likely to happen in the future.

Predictive analysis is used in a wide range of applications, from forecasting sales and demand, to predicting customer behavior, to detecting fraudulent transactions. It involves collecting and analyzing data from a variety of sources, including historical data, customer data, financial data, and social media data, among others.

The process of predictive analysis typically involves the following steps:
1. Defining the problem and identifying the relevant data sources
2. Collecting and cleaning the data
3. Exploring and analyzing the data to identify patterns and trends
4. Selecting an appropriate model or algorithm to use for predictions
5. Training and validating the model using historical data
6. Using the model to make new predictions on new data
7. Monitoring and evaluating the performance of the model over time

Predictive analysis can help organizations make more informed decisions, improve efficiency, and gain a competitive advantage by leveraging insights from data.

It is most commonly used in retail, where workers try to predict which products would be most popular and try to advertise those products as much as possible, and also healthcare, where algorithms analyze patterns and reveal prerequisites for diseases and suggest preventive treatment, predict the results of various treatments and choose the best option for each patient individually, and predict disease outbreaks and epidemics.

# 1. Intro to NumPy and the features it consists

Numpy, by definition, is the fundamental package for scientific computing in Python which can be used to perform mathematical operations, provide multidimensional array objects, and makes data analysis much easier. Numpy is very important and useful when it comes to data analysis, as it can easily use its features to complete and perform any mathematical operation, as well as analyze data files. 

If you don't already have numpy installed, you can do so using ```conda install numpy``` or ```pip install numpy```

Once that is complete, to import numpy in your code, all you must do is:

In [4]:
import numpy as np

# 2. Using NumPy to create arrays
An array is the central data structure of the NumPy library. They are used as containers which are able to store more than one item at the same time. Using the function ```np.array``` is used to create an array, in which you can create multidimensional arrays. 

Shown below is how to create a 1D array:

In [3]:
import numpy  as np # add this line in order to define np and create the array
a = np.array([1, 2, 3])
print(a) 
# this creates a 1D array

[1 2 3]


How could you create a 3D array based on knowing how to make a 1D array?

In [13]:
# create 3D array here
import numpy as np 
b = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(b)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


Arrays can be printed in different ways, especially a more readable format. As we have seen, arrays are printed in rows and columns, but we can change that by using the ```reshape``` function 

In [9]:
c = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(c.reshape(1, 9)) # organizes it all in a single line of output

[[1 2 3 4 5 6 7 8 9]]


In the code segment below, we can also specially select certain rows and columns from the array to further analyze selective data.

In [14]:
print(c[1:, :2])
# the 1: means "start at row 1 and select all the remaining rows"
# the :2 means "select the first two columns"

[[4 5]
 [7 8]]


# 3. Basic array operations

One of the most basic operations that can be performed on arrays is arithmetic operations. With numpy, it is very easy to perform arithmetic operations on arrays. You can add, subtract, multiply and divide arrays, just like you would with regular numbers. When performing these operations, numpy applies the operation element-wise, meaning that it performs the operation on each element in the array separately. This makes it easy to perform operations on large amounts of data quickly and efficiently.

In [15]:
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
print(a + b) # adds each value based on the column the integer is in
print(a - b) # subtracts each value based on the column the integer is in
print(a * b) # multiplies each value based on the column the integer is in
print(a / b) # divides each value based on the column the integer is in

[5 7 9]
[-3 -3 -3]
[ 4 10 18]
[0.25 0.4  0.5 ]


In [16]:
d = np.exp(b)
e = np.sqrt(b)
print(d)
print(e)

[ 54.59815003 148.4131591  403.42879349]
[2.         2.23606798 2.44948974]


From the knowledge of how to use more advanced mathematical expressions than the basic 4 mathematical operations such as exponent and square root, now can you code how to calculate the 3 main trig expressions (sin, cos, tan), natural log, and log10 of a 1D array.

In [19]:
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])

sin= np.sin(a) # calculate sin
print(sin)

cos = np.cos(a) # calculate cos
print(cos)

tan = np.tan(a) # calculate tan
print(tan)

natlog = np.log(a) # calculate natural log
print(natlog)

log10 = np.log10(a) # calculate log10
print(log10)


[0.84147098 0.90929743 0.14112001]
[ 0.54030231 -0.41614684 -0.9899925 ]
[ 1.55740772 -2.18503986 -0.14254654]
[0.         0.69314718 1.09861229]
[0.         0.30103    0.47712125]


# 4. Data analysis using numpy
Numpy provides a convenient and powerful way to perform data analysis tasks on large datasets. One of the most common tasks in data analysis is finding the mean, median, and standard deviation of a dataset. Numpy provides functions to perform these operations quickly and easily. The mean function calculates the average value of the data, while the median function calculates the middle value in the data. The standard deviation function calculates how spread out the data is from the mean. Additionally, numpy provides functions to find the minimum and maximum values in the data. These functions are very useful for gaining insight into the properties of large datasets and can be used for a wide range of data analysis tasks.

In [20]:
data = np.array([2, 5, 12, 13, 19])
print(np.mean(data)) # finds the mean of the dataset
print(np.median(data)) # finds the median of the dataset
print(np.std(data)) # finds the standard deviation of the dataset
print(np.min(data)) # finds the min of the dataset
print(np.max(data)) # finds the max of the dataset

10.2
12.0
6.04648658313239
2
19


Now from learning this, can you find a different way from how we can solve the sum or products of a dataset other than how we learned before?

In [24]:
# create a different way of solving the sum or products of a dataset from what we learned above

print(np.mean(data) * 5) # calculates the sum of the dataset 

print(np.sum(data)) # also calculates the sum of the dataset

print(np.product(data)) # calculates the product of the dataset

51.0
51
29640


Numpy also has the ability to handle CSV files, which are commonly used to store and exchange large datasets. By importing CSV files into numpy arrays, we can easily perform complex operations and analysis on the data, making numpy an essential tool for data scientists and researchers.

```genfromtxt``` and ```loadtxt``` are two functions in the numpy library that can be used to read data from text files, including CSV files.

```genfromtxt``` is a more advanced function that can be used to read text files that have more complex structures, including CSV files. ```genfromtxt``` can handle files that have missing or invalid data, or files that have columns of different data types. It can also be used to skip header lines or to read only specific columns from the file. 

In [27]:
import numpy as np

padres = np.genfromtxt('files/padres.csv', delimiter=',', dtype=str, encoding='utf-8')
# delimiter indicates that the data is separated into columns which is distinguished by commas
# genfromtxt is used to read the csv file itself
# dtype is used to have numpy automatically detect the data type in the csv file

print(padres)

[['Name' ' Position' ' Average' ' HR' ' RBI' ' OPS' ' JerseyNumber']
 ['Manny Machado' ' 3B' ' .298' ' 32' ' 102' ' .897' ' 13']
 ['Fernando Tatis Jr' ' RF' ' .281' ' 42' ' 97' ' .975' ' 23']
 ['Juan Soto' ' LF' ' .242' ' 27' ' 62' ' .853' ' 22']
 ['Xander Bogaerts' ' SS' ' .307' ' 15' ' 73' ' .833' ' 2']
 ['Nelson Cruz' ' DH' ' .234' ' 10' ' 64' ' .651' ' 32']
 ['Matt Carpenter' ' DH' ' .305' ' 15' ' 37' ' 1.138' ' 14']
 ['Jake Cronenworth' ' 1B' ' .239' ' 17' ' 88' ' .722' ' 9']
 ['Ha-Seong Kim' ' 2B' ' .251' ' 11' ' 59' ' .708' ' 7']
 ['Trent Grisham' ' CF' ' .184' ' 17' ' 53' ' .626' ' 1']
 ['Luis Campusano' ' C' ' .250' ' 1' ' 5' ' .593' ' 12']
 ['Austin Nola' ' C' ' .251' ' 4' ' 40' ' .649' ' 26']
 ['Jose Azocar' ' OF' ' .257' ' 0' ' 10' ' .630' ' 28']]


```loadtxt``` is a simpler function that can be used to read simple text files that have a regular structure, such as files that have only one type of data (such as all integers or all floats). ```loadtxt``` can be faster than ```genfromtxt``` because it assumes that the data in the file is well-structured and can be easily parsed.

In [28]:
import numpy as np

padres = np.loadtxt('files/padres.csv', delimiter=',', dtype=str, encoding='utf-8')
print(padres)

[['Name' ' Position' ' Average' ' HR' ' RBI' ' OPS' ' JerseyNumber']
 ['Manny Machado' ' 3B' ' .298' ' 32' ' 102' ' .897' ' 13']
 ['Fernando Tatis Jr' ' RF' ' .281' ' 42' ' 97' ' .975' ' 23']
 ['Juan Soto' ' LF' ' .242' ' 27' ' 62' ' .853' ' 22']
 ['Xander Bogaerts' ' SS' ' .307' ' 15' ' 73' ' .833' ' 2']
 ['Nelson Cruz' ' DH' ' .234' ' 10' ' 64' ' .651' ' 32']
 ['Matt Carpenter' ' DH' ' .305' ' 15' ' 37' ' 1.138' ' 14']
 ['Jake Cronenworth' ' 1B' ' .239' ' 17' ' 88' ' .722' ' 9']
 ['Ha-Seong Kim' ' 2B' ' .251' ' 11' ' 59' ' .708' ' 7']
 ['Trent Grisham' ' CF' ' .184' ' 17' ' 53' ' .626' ' 1']
 ['Luis Campusano' ' C' ' .250' ' 1' ' 5' ' .593' ' 12']
 ['Austin Nola' ' C' ' .251' ' 4' ' 40' ' .649' ' 26']
 ['Jose Azocar' ' OF' ' .257' ' 0' ' 10' ' .630' ' 28']]


In [29]:
for i in padres:
    print(",".join(i))

Name, Position, Average, HR, RBI, OPS, JerseyNumber
Manny Machado, 3B, .298, 32, 102, .897, 13
Fernando Tatis Jr, RF, .281, 42, 97, .975, 23
Juan Soto, LF, .242, 27, 62, .853, 22
Xander Bogaerts, SS, .307, 15, 73, .833, 2
Nelson Cruz, DH, .234, 10, 64, .651, 32
Matt Carpenter, DH, .305, 15, 37, 1.138, 14
Jake Cronenworth, 1B, .239, 17, 88, .722, 9
Ha-Seong Kim, 2B, .251, 11, 59, .708, 7
Trent Grisham, CF, .184, 17, 53, .626, 1
Luis Campusano, C, .250, 1, 5, .593, 12
Austin Nola, C, .251, 4, 40, .649, 26
Jose Azocar, OF, .257, 0, 10, .630, 28


# Pandas
### What is Pandas
Pandas is a Python library used for working with data sets. A python library is something  It has functions for analyzing, cleaning, exploring, and manipulating data.

### Why Use Pandas?
Pandas allows us to analyze big data and make conclusions based on statistical theories. Pandas can clean messy data sets, and make them readable and relevant. Also it is a part of data analysis, and data manipulation.

### What Can Pandas Do?
Pandas gives you answers about the data. Like:
- Is there a correlation between two or more columns?
- What is average value
- Max value
- Min value
- How to load data 
- Delete data 
- Sort Data.

Pandas are also able to delete rows that are not relevant, or contains wrong values, like empty or NULL values. This is called cleaning the data.

# Basics of Pandas.

In [30]:
import pandas as pd # how to bring the panda library into your code
# What this does is it calls the python pandas library and this code segment is needed whenever incorporating pandas.

#### DICTIONARIES AND DATASETS
- One way you are able to manipulate a pandas data set is by creating a dictionary and calling it as seen with the dict data 1 and pd.dataframe which is a way to print the set.

In [31]:
import pandas as pd

data1 = {
  'teams': ["BARCA", "REAL", "ATLETICO"],
  'standings': [1, 2, 3]
}

myvar = pd.DataFrame(data1)

print(myvar)


      teams  standings
0     BARCA          1
1      REAL          2
2  ATLETICO          3


### Indexing and manipulaton of data through lists.
- With pandas you can also organize the data which is one of its biggest perks, we call this indexing, this is when we define the first column in a data frame.

In [32]:
# Here is an example using lists and an index.
import pandas as pd 

score = [5/5, 5/5, 1/5]

myvar = pd.Series(score, index = ["math", "science", "pe"])

print(myvar)

math       1.0
science    1.0
pe         0.2
dtype: float64


# Pandas Classes 
Within pandas the library consits of a lot of functions which allow you to manipulate datasets in lists dictionsaries and csv files here are some of the ones we are going to cover (hint: take notes on these)
- Series
- Index
- PeriodIndex: allows to repeat data over time
- DataframeGroupedBy: allows to organize data and calculate different functions
- Categorical: sets up a category for something and puts it within the categories and allows for better orginzation
- Time Stamp: displays a single time


#  PeriodIndex 
- This allows for a way to repeat data over time that it occurs as seen from january 2022 to december 2023. You can use Y for years, M for months, and D for days.


In [33]:
import pandas as pd


time = pd.period_range('2022-01', '2022-12', freq='M')


print(time)

PeriodIndex(['2022-01', '2022-02', '2022-03', '2022-04', '2022-05', '2022-06',
             '2022-07', '2022-08', '2022-09', '2022-10', '2022-11', '2022-12'],
            dtype='period[M]')


Now implement a way to show a period index from June 2022 to July 2023 in days.

In [34]:
# use period index to show - in days - June 2022 to 2023

import pandas as pd
time = pd.period_range('2022-06', '2023-06', freq='D')
print(time)

PeriodIndex(['2022-06-01', '2022-06-02', '2022-06-03', '2022-06-04',
             '2022-06-05', '2022-06-06', '2022-06-07', '2022-06-08',
             '2022-06-09', '2022-06-10',
             ...
             '2023-05-23', '2023-05-24', '2023-05-25', '2023-05-26',
             '2023-05-27', '2023-05-28', '2023-05-29', '2023-05-30',
             '2023-05-31', '2023-06-01'],
            dtype='period[D]', length=366)


# Dataframe Grouped By 
- This allows for you to organize your data and calculate the different functions such as
- count(): returns the number of non-null values in each group.
- sum(): returns the sum of values in each group.
- mean(): returns the mean of values in each group.
- min(): returns the minimum value in each group.
- max(): returns the maximum value in each group.
- median(): returns the median of values in each group.
- var(): returns the variance of values in each group.
- agg(): applies one or more functions to each group and returns a new DataFrame with the results.


In [35]:
import pandas as pd

data = {
    'Category': ['E', 'F', 'E', 'F', 'E', 'F', 'E', 'F'],
    'Value': [100, 250, 156, 255, 240, 303, 253, 3014]
}
df = pd.DataFrame(data)


grouped = df.groupby('Category') #GUESS WHAT THIS WOULD BE IF WE WERE LOOKING FOR COMBINED TOTALS!()

print(grouped)


### Categorical 
- This sets up a category for something and puts it within the categories and allows for better orginzation 

In [37]:
import pandas as pd

colors = pd.Categorical(['yellow', 'orange', 'blue', 'yellow', 'orange'], categories=['yellow', 'orange', 'blue'])

print(colors)


['yellow', 'orange', 'blue', 'yellow', 'orange']
Categories (3, object): ['yellow', 'orange', 'blue']


### Timestamp Class
- This allows to display a single time which can be useful when working with datasets that deal with time allowing you to manipulate the time you do something and how you do it. 

In [38]:
import pandas as pd


timing = pd.Timestamp('2023-02-05 02:00:00')

print(timing) # this would print the timestamp of febrary 5th, 2023, at 02:00:00


2023-02-05 02:00:00


# CSV FILES!
- A csv file contains data and within pandas you are able to call the function and you are able to manipulate the data with the certain data classes talked about above. 

- Name, Position, Average, HR, RBI, OPS, JerseyNumber
- Manny Machado, 3B, .298, 32, 102, .897, 13
- Tatis Jr, RF, .281, 42, 97, .975, 23
- Juan Soto, LF, .242, 27, 62, .853, 22
- Xanger Bogaerts, SS, .307, 15, 73, .833, 2
- Nelson Cruz, DH, .234, 10, 64, .651, 32
- Matt Carpenter, DH, .305, 15, 37, 1.138, 14
- Cronezone, 1B, .239, 17, 88, .722, 9
- Ha-Seong Kim, 2B, .251, 11, 59, .708, 7
- Trent Grisham, CF, .184, 17, 53, .626, 1
- Luis Campusano, C, .250, 1, 5, .593, 12
- Austin Nola, C, .251, 4, 40, .649, 26
- Jose Azocar, OF, .257, 0, 10, .630, 28

QUESTION: WHAT DO YOU GUYS THINK THE INDEX FOR THIS WOULD BE? |
ANSWER: 7

**Can you explain what is going on in this code segment below? (hint: define what ascending= false means, and df. head means)**

- First, you read through the padres csv file, and sort the data by the player's first names in the opposite of an alphabetical order (ascending= false means that we want the names printed in reverse alphabetical order)
- df.head means that we want to get the top 10 players
- df. tail means that we want to get the bottom 10 players

In [39]:
import pandas as pd

#read csv and sort 'Duration' largest to smallest
df = pd.read_csv('files/padres.csv').sort_values(by=['Name'], ascending=False)

print("--Duration Top 10---------")
print(df.head(10))

print("--Duration Bottom 10------")
print(df.tail(10))
print(', '.join(df.tail(10)))

--Duration Top 10---------
                Name  Position   Average   HR   RBI    OPS   JerseyNumber
3    Xander Bogaerts        SS     0.307   15    73  0.833              2
8      Trent Grisham        CF     0.184   17    53  0.626              1
4        Nelson Cruz        DH     0.234   10    64  0.651             32
5     Matt Carpenter        DH     0.305   15    37  1.138             14
0      Manny Machado        3B     0.298   32   102  0.897             13
9     Luis Campusano         C     0.250    1     5  0.593             12
2          Juan Soto        LF     0.242   27    62  0.853             22
11       Jose Azocar        OF     0.257    0    10  0.630             28
6   Jake Cronenworth        1B     0.239   17    88  0.722              9
7       Ha-Seong Kim        2B     0.251   11    59  0.708              7
--Duration Bottom 10------
                 Name  Position   Average   HR   RBI    OPS   JerseyNumber
4         Nelson Cruz        DH     0.234   10    64  0.6

In [40]:
import pandas as pd


df = pd.read_csv("./files/housing.csv")


mode_total_rooms = df['total_rooms'].mode()


print(f"The mode of the 'total_rooms' column is: {mode_total_rooms}")

The mode of the 'total_rooms' column is: 0    1527.0
Name: total_rooms, dtype: float64


In [41]:
import pandas as pd

df = pd.read_csv("./files/housing.csv")


grouped_df = df.groupby('total_rooms')


agg_df = grouped_df.agg({'total_rooms': 'sum', 'population': 'mean', 'longitude': 'count'})

# WHAT DO YOU GUYS THINK df.agg means in context of pandas and what does it stand for.


print(agg_df)


             total_rooms  population  longitude
total_rooms                                    
2.0                  2.0         6.0          1
6.0                  6.0         8.0          1
8.0                  8.0        13.0          1
11.0                11.0        24.0          1
12.0                12.0        18.0          1
...                  ...         ...        ...
30450.0          30450.0      9419.0          1
32054.0          32054.0     15507.0          1
32627.0          32627.0     28566.0          1
37937.0          37937.0     16122.0          1
39320.0          39320.0     16305.0          1

[5926 rows x 3 columns]


# Our Frontend Data Analysis Project
[Link](https://paravsalaniwal.github.io/T3Project/DataAnalysisProject/)

# Popcorn Hacks
- Complete fill in the blanks for Predictive Analysis Numpy
- Takes notes on Panda where it asks you to
- Complete code segment tasks in Panda and Numpy

# Main Hack
- Make a data file - content is up to you, just make sure there are integer values - and print
- Run Panda and Numpy commands
    - Panda:
        - Find Min and Max values
        - Sort in order - can be order of least to greatest or vice versa
        - Create a smaller dataframe and merge it with your data file
    - Numpy:
        -  Random number generation
        - create a multi-dimensional array (multiple elements)
        - create an array with linearly spaced intervals between values


## Make a Data File

In [15]:
import pandas as pd

# Read CSV file into a pandas dataframe
df = pd.read_csv('./files/carspecs.csv')

# Print the dataframe
print(df)

                modelName        carType  mileage  seatingCapacity
0        Chevrolet Blazer            SUV       25                5
1             Ford Escape            SUV       22                5
2       Chrysler Pacifica        Minivan       40                7
3              Ford F-150   Pickup Truck       17                5
4           Toyota Sienna        Minivan       36                8
5        Dodge Challenger     Muscle Car       13                2
6      Nissan Skyline R35     Sports Car       18                4
7   Chevrolet Corvette C8     Sports Car       19                2
8             Honda Pilot            SUV       22                8
9             Honda Civic          Sedan       35                5
10         Toyota Corolla          Sedan       33                5
11           Toyota Camry          Sedan       32                5
12           Ford Mustang     Sports Car       24                4
13       Chevrolet Camaro     Sports Car       24             

## Find Max & Min Values

In [17]:
import pandas as pd

data = pd.read_csv('files/carspecs.csv').sort_values(by=['mileage'], ascending=False)
columns = data[['modelName','carType','mileage', 'seatingCapacity']]

max_mileage = df['mileage'].max()
print(f"The maximum mileage of the cars in the list is: {max_mileage}")
print("Here are the details of the car with the most mileage:")
print(columns[columns.mileage == columns.mileage.max()])
print()

min_mileage = df['mileage'].min()
print(f"The minimum mileage of the cars in the list is: {min_mileage}")
print("Here are the details of the car with the least mileage:")
print(columns[columns.mileage == columns.mileage.min()])
print()

The maximum mileage of the cars in the list is: 40
Here are the details of the car with the most mileage:
           modelName   carType  mileage  seatingCapacity
2  Chrysler Pacifica   Minivan       40                7

The minimum mileage of the cars in the list is: 13
Here are the details of the car with the least mileage:
          modelName      carType  mileage  seatingCapacity
5  Dodge Challenger   Muscle Car       13                2



## Sort in Order

In [18]:
import pandas as pd

#read csv and sort 'Duration' largest to smallest
df = pd.read_csv('files/carspecs.csv').sort_values(by=['mileage'], ascending=False)

print("--Top 5 Mileage---------")
print(df.head(5))

print("--Bottom 5 Mileage------")
print(df.tail(5))

--Top 5 Mileage---------
            modelName   carType  mileage  seatingCapacity
2   Chrysler Pacifica   Minivan       40                7
4       Toyota Sienna   Minivan       36                8
9         Honda Civic     Sedan       35                5
10     Toyota Corolla     Sedan       33                5
11       Toyota Camry     Sedan       32                5
--Bottom 5 Mileage------
               modelName        carType  mileage  seatingCapacity
8            Honda Pilot            SUV       22                8
7  Chevrolet Corvette C8     Sports Car       19                2
6     Nissan Skyline R35     Sports Car       18                4
3             Ford F-150   Pickup Truck       17                5
5       Dodge Challenger     Muscle Car       13                2


## Create Small Data Frame & Merge

In [39]:
import pandas as pd

# create a new DataFrame with additional data
new_data = pd.DataFrame({
    'modelName': ['Jeep Wrangler', 'Mazda MX-5 Miata', 'Audi A4'],
    'carType': ['SUV', 'Sports Car', 'Sedan'],
    'mileage': [25, 30, 27],
    'seatingCapacity': [5, 2, 5]
})

# load the original CSV data file into a DataFrame
original_data = pd.read_csv('files/carspecs.csv')

# concatenate the original DataFrame with the new DataFrame
combined_data = pd.concat([original_data, new_data])

# write the combined DataFrame to a new CSV file
combined_data.to_csv('files/carspecs.csv', index=False)

print('Old Data Frame:')
print(df)
print('New Data Frame:')
print(combined_data)

Old Data Frame:
                modelName        carType  mileage  seatingCapacity
0        Chevrolet Blazer            SUV       25                5
1             Ford Escape            SUV       22                5
2       Chrysler Pacifica        Minivan       40                7
3              Ford F-150   Pickup Truck       17                5
4           Toyota Sienna        Minivan       36                8
5        Dodge Challenger     Muscle Car       13                2
6      Nissan Skyline R35     Sports Car       18                4
7   Chevrolet Corvette C8     Sports Car       19                2
8             Honda Pilot            SUV       22                8
9             Honda Civic          Sedan       35                5
10         Toyota Corolla          Sedan       33                5
11           Toyota Camry          Sedan       32                5
12           Ford Mustang     Sports Car       24                4
13       Chevrolet Camaro     Sports Car      

## Random Number Generation

In [31]:
import numpy as np

randomNumber = np.random.randint(1, 1000)

print(randomNumber)

96


## Multi-Dimensional Array

In [33]:
# multi-dimensional array
import numpy as np 
carspecs = np.array([['Model Name', 'Car Type', 'Mileage', 'seatingCapacity'], ['Chevrolet Blazer', 'SUV', '25', '5'], ['Ford Escape', 'SUV', '22', '5'], ['Chrysler Pacifica', 'Miniva', '40', '7'], ['Ford F-150', 'Pickup Truck', '17', '5'], ['Toyota Sienna', 'Minivan', '36', '8'], ['Dodge Challenger', 'Muscle Car', '13', '2'], ['Nissan Skyline R35r', 'Sports Car', '18', '4'], ['Chevrolet Corvette C8', 'Sports Car', '19', '2'], ['Honda Pilot', 'SUV', '22', '8']])
print(carspecs)

[['Model Name' 'Car Type' 'Mileage' 'seatingCapacity']
 ['Chevrolet Blazer' 'SUV' '25' '5']
 ['Ford Escape' 'SUV' '22' '5']
 ['Chrysler Pacifica' 'Miniva' '40' '7']
 ['Ford F-150' 'Pickup Truck' '17' '5']
 ['Toyota Sienna' 'Minivan' '36' '8']
 ['Dodge Challenger' 'Muscle Car' '13' '2']
 ['Nissan Skyline R35r' 'Sports Car' '18' '4']
 ['Chevrolet Corvette C8' 'Sports Car' '19' '2']
 ['Honda Pilot' 'SUV' '22' '8']]


## Array w/ Linearly Spaced Intervals

In [42]:
import numpy as np

# Create a 2D array with 4 rows and 6 columns, where each element is linearly spaced between 0 and 1
arr = np.linspace(0, 1, num=24).reshape(4, 6)
print(arr)

[[0.         0.04347826 0.08695652 0.13043478 0.17391304 0.2173913 ]
 [0.26086957 0.30434783 0.34782609 0.39130435 0.43478261 0.47826087]
 [0.52173913 0.56521739 0.60869565 0.65217391 0.69565217 0.73913043]
 [0.7826087  0.82608696 0.86956522 0.91304348 0.95652174 1.        ]]
